### PR TITLE
feat: delaying cloud mode events based on flag provided in options

### PR DESCRIPTION
**Pull request by mihir-4116**
## PR Description

Please include a summary of the change along with the relevant motivation and context.

## Notion ticket

Ticket link

## Screenshots

Please add screenshots for any new features or UI bug fixes for the following browsers -

- Chrome
  >
- Firefox
  >
- Safari
  >

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: src/core/analytics.js [hunk 1]
```diff
@@ -789,19 +807,26 @@ class Analytics {
...
-      transformToServerNames(rudderElement.message.integrations);
-      rudderElement.message.integrations = getMergedClientSuppliedIntegrations(
-        this.integrationsData,
-        clientSuppliedIntegrations,
-      );
+      // Holding the cloud mode events based on flag and integrations load check
+      if (
+        !options?.bufferDataPlaneEventsUntilReady ||
+        (this.clientIntegrationObjects &&
+          this.clientIntegrationObjects.every((intg) => !intg.isReady || intg.isReady()))
```
Comment (saikumarrs) [inline src/core/analytics.js:814]: Just use the logic on Ln 779.

File: src/core/analytics.js [hunk 2]
```diff
@@ -789,19 +807,26 @@ class Analytics {
...
 
-      // convert integrations object to server identified names, kind of hack now!
-      transformToServerNames(rudderElement.message.integrations);
-      rudderElement.message.integrations = getMergedClientSuppliedIntegrations(
-        this.integrationsData,
-        clientSuppliedIntegrations,
-      );
+      // Holding the cloud mode events based on flag and integrations load check
+      if (
+        !options?.bufferDataPlaneEventsUntilReady ||
```
Comment (saikumarrs) [inline src/core/analytics.js:812]: This should be a load option and not an event-level option.

**Review by saikumarrs — CHANGES_REQUESTED** [review]
This is a feature not a fix.

File: src/core/analytics.js (hunk 1 above)
Comment (mihir-4116) [reply src/core/analytics.js:814]: updated

File: src/core/analytics.js (hunk 2 above)
Comment (mihir-4116) [reply src/core/analytics.js:812]: updated

File: src/core/analytics.js [hunk 3]
```diff
@@ -789,19 +806,22 @@ class Analytics {
...
       }
 
-      // convert integrations object to server identified names, kind of hack now!
-      transformToServerNames(rudderElement.message.integrations);
-      rudderElement.message.integrations = getMergedClientSuppliedIntegrations(
-        this.integrationsData,
-        clientSuppliedIntegrations,
-      );
+      // Holding the cloud mode events based on flag and integrations load check
+      if (!this.options?.bufferDataPlaneEventsUntilReady || this.clientIntegrationObjects) {
```
Comment (saikumarrs) [inline src/core/analytics.js:810]: Sanitizing this option is required. Refer to Ln 1092-1094.

**Review by saikumarrs — COMMENTED** [review]
LGTM

File: src/core/analytics.js (hunk 3 above)
Comment (mihir-4116) [reply src/core/analytics.js]: updated

File: src/core/analytics.js [hunk 4]
```diff
@@ -1069,6 +1090,10 @@ class Analytics {
       this.loadIntegration = !!options.loadIntegration;
     }
 
+    if (options && options.bufferDataPlaneEventsUntilReady != undefined) {
+      this.bufferDataPlaneEventsUntilReady = !!options.bufferDataPlaneEventsUntilReady;
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply src/core/analytics.js]: updated

**Review by saikumarrs — APPROVED** [review]

**Comment by bardisg** [comment]
What is this? This is not integration related change. This should be owned by SDK team. As it stands I would like to have a meeting with the requirement as provided by PO in order to consider

**Comment by mihir-4116** [comment]
closing this pr as new one created [here](https://github.com/rudderlabs/rudder-sdk-js/pull/1038)

**Pull request closed without merging**